### DISCUSSION TITLE
Auth User extra fields

### DISCUSSION THREAD

**Discussion by nandox5** [supabase/supabase#8893] [Questions]
Is there a way to add more data fields to the user? I need to be able to add custom roles/permission field along with other pieces of information. 

OR can we create a relation from a table to the authentication?

**Comment by NickRTR** [comment]
Hello,

You can add additional user data to user_metadata in the user auth object from supabase. 

https://supabase.com/docs/guides/auth/managing-user-data

**Comment by nandox5** [comment]
Forgot to mention that i am using prisma using their schema system. Is it possible with prisma to relate the two? (Profile table with users table in auth)

**Reply by NickRTR** [reply]
Oh, I don't know how this works with prisma, I'm sorry. But basically the user in supabase is just a table with many columns. You can even access the table from your Supabase Dashboard: 
> table editor > select schema > "auth" > users

Maybe that is some advice for you.

**Comment by bennik88** [comment]
It's a bit of a design decision on your side. Here are a few options you are having for saving additional user information or roles.

**For user roles you can use:**
- Add roles and permissions with Postgres Row level security, tables and custom types. Depending on how much SQL knowledge you already have this should help you to get started. On the linked page you find the Postgresql source code at the bottom which is showing all the logic behind it. If you take your time to read it, it should help you to get a better understanding of how to set up roles and permissions and link them with relationships in your tables to later secure them with Row level security. A little bit advanced as a starter. Example with [Slack Clone](https://github.com/supabase/examples/tree/main/supabase-js-v1/slack-clone/nextjs-slack-clone)
- [Supabase Custom Claims](https://github.com/supabase-community/supabase-custom-claims) which allow you to set claims like `superadmin`, `admin`, `manager` or whatever roles you have in mind. This is very similar to a Firebase like approach. This could potentially be a simpler solution as it doesn't involve setting up related tables. But in its basic form, it lacks the control of permissions. Yet, custom claims can be very powerful by themself.

**For saving additional user information**
- You can use user_metadata on sign up for example it could look like this depending on your front-end language:
```js
const { user, session, error } = await supabase.auth.signUp(
  {
    email: 'example@email.com',
    password: 'example-password',
  },
  {
    data: {
      first_name: 'John',
      age: 27,
    },
  }
)
```
[Docs to this example](https://supabase.com/docs/reference/javascript/auth-signup)

- An alternative way is to use a profiles table and link the profile primary key `id` it with a foreign key to your auth.users table `id`.
- You can also use both together which is very commonly used together with triggers. An example to combine these two methods [here](https://database.dev/handle-new-user).

I hope this helps. It's a little bit of work to get the hang out of it, but at the end it's worth it ;)

**Comment by nandox5** [comment]
Thanks for the info. I understand what i have to do now, the extra work would not be an issue here. :)

The second issue here is the orm that i am using. Looking at the docs from supabase, i didnt see an example on how i can access the other schemas (such as auth) through prisma. Is this a supported case with prisma and supabase?

Im leaning towards the second method where i create a profile table and link the users table with a foreign  key. Hopefully it’s doable with prisma.

**Comment by nandox5** [comment]
has anyone been able to access the other schemes with prisma? Auth in particular.
